### PR TITLE
Revise TAP 3 (Multi-role delegations)

### PR DESCRIPTION
**Pull request by vladimir-v-diaz**
We decided to amend [TAP 3](https://github.com/theupdateframework/taps/blob/master/tap3.md) to include new metadata fields (e.g., min_roles_in_agreement), convert the "names" field to a list so that delegations can be ordered, etc.  Please visit the [issue discussion for TAP 3](https://github.com/theupdateframework/taps/issues/32) for the full discussion of the proposed revisions to TAP 3.

@awwad Can you please review the revisions and verify that the text is clear?

### CODE REVIEW

File: tap3.md [hunk 1]
```diff
@@ -96,68 +94,88 @@ the given set of targets.
 }
 </pre>
 
-## The new targets metadata file format
+## The new file format of targets metadata
```
Comment (awwad) [inline tap3.md]: You provide two files in the new format:
- a generic format definition with some examples in it
- a better example

I think this is confusing in a reference document and also generally to a casual reader. I think the format definition should be entirely generic, with one delegation listed and a snipped second (just to indicate that there's a list of delegations), and no specific values, just labels. Then you can move on to the better example you have below.

I can PR or commit that if you don't mind.

You may even want to exclude most of the comments in the definition and keep them in the example.

File: tap3.md [hunk 2]
```diff
@@ -174,54 +192,69 @@ targets:
...
         {
           // These targets must be signed by this <b>single</b> role.
+          <b>"name": "first_delegation",</b>
           "paths": ["/foo/*.pkg"],
-          <b>"names"</b>: <b>{</b>
+          "terminating": false,
+          <b>"min_roles_in_agreement": 1,</b>
+          "roleinfo": [
+            {
             // This role must sign them using all 3 of these keys.
```
Comment (awwad) [inline tap3.md]: I suggest cutting this line and adding some description 8 lines up, in the delegation description, explaining the full meaning of this delegation there: "This first delegation requires one role ("min_roles_in_agreement" = 1), "alice", to sign targets matching pattern "/foo/*.pkg". It also specifies that the "alice" role must be signed with all three of the keyids listed ("threshold" = 3) in order for to be validated."


I guess you could alternatively move this comment line down three lines into "keyids" and write instead:
'These three keys are being associated with this role. Per "threshold", all three must be used to sign this role. Per "min_roles_in_agreement" and "roleinfo", only this role, "alice", need sign files matching /foo/*.pkg in order for those files to be validated.'

I prefer the first.

File: tap3.md [hunk 3]
```diff
@@ -174,54 +192,69 @@ targets:
...
                 "de7d1b721a1e0632b7cf04edf5032c8ecffa9f9a08492152b926f1a5a7e765d7"
               ],
               "threshold": 3
-            <b>}</b>
-          <b>}</b>,
-          ...
+            }
+          ]
         },
         // This is the second delegation.
```
Comment (awwad) [inline tap3.md]: I suggest you crop some of the comments below and write this here:

This, second, delegation requires that both roles ("min_roles_in_agreement" = 2) 'release-engineering' and 'quality-assurance' sign for the target files in order to validate them via this delegation. 'release-engineering' is expected to be signed with the single listed key for that role ("threshold" = 1). 'quality-assurance' must be signed with at least two of the three listed keys for that role ("threshold" = 2).

File: tap3.md [hunk 4]
```diff
@@ -86,7 +86,7 @@ sign the given set of targets.
           "terminating": false,
         }
         // Note that, unfortunately, there is no way to require <b>multiple</b>
-        // roles to sign targets in a single delegation.
+        // roles as required to sign targets in a single delegation.
```
Comment (awwad) [inline tap3.md]: this one is redundant now. "require multiple roles as required"

don't need the change

**Comment by vladimir-v-diaz** [comment]
Thanks for the review, Seb.

**Pull request merged**